src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                        # then we can just stop.
+                        if taskloop_positions[j] >= next_dependence:
+                            break
+                        # If the jth taskloop has no dependency then continue
+                        if dependence_position[j] is None:
+                            continue
+                        # Check if next_dependence will satisfy the jth
+                        # taskloops dependency
+                        if next_dependence <= dependence_position[j]:
+                            dependence_position[j] = None
+                            continue
+                        # Check if the jth taskloops dependence will satisfy
+                        # the next_dependence
+                        if dependence_position[j] < next_dependence:
+                            dependence_position[i] = None
+                            # If it does then we can move to the next taskloop
+                            break
+            # dependence_position now contains only the required dependencies
+            # to satisfy the full superset of dependencies. We can loop over
+            # these by index, and if dependence_position[i] is not None then
+            # go to its forward dependency, and insert a TaskwaitDirective
+            # immediately before.
+            for i, dep_pos in enumerate(dependence_position):
+                if dep_pos is not None:
+                    forward_dep = OMPTaskwaitTrans.get_forward_dependence(
+                            taskloops[i], node)
+                    fdep_parent = forward_dep.parent
+                    # Find the position of the forward_dep in its parent's
+                    # children list
+                    loc = -1
+                    for j, child in enumerate(fdep_parent.children):
+                        if child is forward_dep:
+                            loc = j
+                            break
+                    # We've found the position, so we now insert an
+                    # OMPTaskwaitDirective in that location instead
+                    fdep_parent.addchild(OMPTaskwaitDirective(), loc)
+            if create_endtaskwait:
+                create_endtaskwait = False
+                # For each taskloop that needed a taskwait at the end
+                for taskloop in endwaits:
+                    unsatisfied = True
+                    taskloop_pos = taskloop.abs_position
+                    # Find all the taskwaits in the task_region
+                    nodes = task_region.walk(OMPTaskwaitDirective)
+                    for node1 in nodes:
+                        # If the taskwait appears after this taskloop then
+                        # its dependency is satisfied
+                        if node1.abs_position > taskloop_pos:
+                            unsatisfied = False
+                    create_endtaskwait = create_endtaskwait or unsatisfied
+                if create_endtaskwait:
```

Comment:
If I've understood this, as soon as we have `unsatisfied = True` then we know we need to create an OMPTaskWaitDirective at the end of this region? That being so I think we could simplify a bit:

    if any(node1.abs_position <= taskloop_pos for node1 in nodes):
        task_region.directive_body.addchild(....)
        break

(but I may be wrong)

Reply:
Yes, I think thats almost correct, I think its:
```
if all(node1.abs_position <= taskloop_pos for node1 in nodes):
    # do stuff and break
```
as its only if there is no taskwait between this taskloop and the end of the single that we need the extra barrier.